keys.go

```diff
@@ -50,116 +74,199 @@ func findKey(session *pkcs11Session, id []byte, label []byte, keyclass *uint, ke
+	if err != nil {
+		return nil, err
+	}
+	if pubHandle == nil {
+		// We can't return a Signer if we don't have private and public key. Treat it as an error.
+		return nil, errors.New("could not find public key to match private key")
+	}
+
+	var pub crypto.PublicKey
+	switch keyType {
+	case pkcs11.CKK_DSA:
+		if pub, err = exportDSAPublicKey(session, *pubHandle); err != nil {
+			return nil, err
+		}
+		return &pkcs11PrivateKeyDSA{
+			pkcs11PrivateKey: pkcs11PrivateKey{
+				pkcs11Object: pkcs11Object{
+					handle:  *privHandle,
+					context: c,
+				},
+				pubKeyHandle: *pubHandle,
+				pubKey:       pub,
+			}}, nil
+
+	case pkcs11.CKK_RSA:
+		if pub, err = exportRSAPublicKey(session, *pubHandle); err != nil {
+			return nil, err
+		}
+		return &pkcs11PrivateKeyRSA{
+			pkcs11PrivateKey: pkcs11PrivateKey{
+				pkcs11Object: pkcs11Object{
+					handle:  *privHandle,
+					context: c,
+				},
+				pubKeyHandle: *pubHandle,
+				pubKey:       pub,
+			}}, nil
+
+	case pkcs11.CKK_ECDSA:
+		if pub, err = exportECDSAPublicKey(session, *pubHandle); err != nil {
+			return nil, err
+		}
+		return &pkcs11PrivateKeyECDSA{
+			pkcs11PrivateKey: pkcs11PrivateKey{
+				pkcs11Object: pkcs11Object{
+					handle:  *privHandle,
+					context: c,
+				},
+				pubKeyHandle: *pubHandle,
+				pubKey:       pub,
+			}}, nil
+
+	default:
+		return nil, errors.Errorf("unsupported key type: %X", keyType)
+	}
+}
+
 // FindKeyPair retrieves a previously created asymmetric key pair, or nil if it cannot be found.
 //
 // At least one of id and label must be specified. If the private key is found, but the public key is
 // not, an error is returned because we cannot implement crypto.Signer without the public key.
 func (c *Context) FindKeyPair(id []byte, label []byte) (Signer, error) {
+	result, err := c.FindKeyPairs(id, label)
+	if err != nil {
+		return nil, err
+	}
 
-	if c.closed.Get() {
-		return nil, errClosed
+	if len(result) == 0 {
+		return nil, nil
 	}
 
-	var k Signer
+	return result[0], nil
+}
 
-	err := c.withSession(func(session *pkcs11Session) error {
+// FindKeyPairs retrieves all matching asymmetric key pairs, or a nil slice if none can be found.
+//
+// At least one of id and label must be specified.
+// If a private key is found, but the corresponding public key is not, the key is not returned because we cannot
+// implement crypto.Signer without the public key.
+func (c *Context) FindKeyPairs(id []byte, label []byte) ([]Signer, error) {
+	if id == nil && label == nil {
+		return nil, errors.New("id and label cannot both be nil")
+	}
 
-		var pub crypto.PublicKey
+	attributes := NewAttributeSet()
 
-		privHandle, err := findKey(session, id, label, uintPtr(pkcs11.CKO_PRIVATE_KEY), nil)
-		if err != nil {
-			return err
-		}
-		if privHandle == nil {
-			// Cannot continue, no key found
-			return nil
-		}
+	if id != nil {
+		attributes[CkaId] = pkcs11.NewAttribute(CkaId, id)
```

Comment:
Ah, I thought the `Set` interface was for external use to prevent a panic if the user passes an invalid type, whereas I know the type I'm passing is valid so thought I could avoid the error check. I'll change them to use `Set` instead.